File: 08_tf2_tutorial/src/lookup_transform_python/lookup_transform_python/lookup_transform.py
import rclpy
from rclpy.node import Node
from rclpy.time import Time

from tf2_ros import TransformException
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener

class LookupTransform(Node):
    def __init__(self):
        # 初始化节点，节点名为lookup_transform_python
        super().__init__('lookup_transform_python')

        # 创建Buffer
        self.tf_buffer = Buffer(node=self)

        # 创建TransformListener
        self.tf_listener = TransformListener(
            self.tf_buffer,
            self,
            spin_thread=False
        )

        # 创建定时器
        # 每 1s 查询一次Transform，即查询频率约为1 Hz
        self.timer = self.create_timer(
            1.0,
            self.timer_callback
        )

        # 打印启动日志
        self.get_logger().info(
            '开始查询Transform: odom -> base_link'
        )

    def _lookup_latest_transform(self, target, source):
        try:
            # 查询最新可用的 Transform
            # target_frame 为 target
            # source_frame 为 source
            transform = self.tf_buffer.lookup_transform(
                target,
                source,
                Time()
            )

        except TransformException as ex:
            # 当前Buffer中不存在可用Transform时打印警告
            self.get_logger().warn(
                f'暂时无法查询 {source} 转换到 {target} 的 Transform: {ex}'
            )
            return

        # 获取平移分量
        translation = transform.transform.translation

        # 获取旋转分量
        rotation = transform.transform.rotation

        # 打印查询结果
        self.get_logger().info(
            f'查询到 {source} 转换到 {target} Transform: '
            f'translation=[{translation.x:.3f}, '
            f'{translation.y:.3f}, '
            f'{translation.z:.3f}], '
            f'rotation=[{rotation.x:.3f}, '
            f'{rotation.y:.3f}, '
            f'{rotation.z:.3f}, '
            f'{rotation.w:.3f}]'
        )
    
    def timer_callback(self):
        self.get_logger().info("=" * 20)
        # 查询 base_link(source) 转换到 odom(target) 的 Transform
        self._lookup_latest_transform('odom', 'base_link')

        # 查询 odom 转换到 base_link 的 Transform
        # Buffer 帮我们完成逆变换
        self._lookup_latest_transform('base_link', 'odom')

        # 查询 base_link 转换到 map 的 Transform
        # Buffer 帮我们完成复合变换
        self._lookup_latest_transform('map', 'base_link')      
        self.get_logger().info("=" * 20)

def main(args=None):
    # 初始化ROS 2
    rclpy.init(args=args)

    # 创建节点
    node = LookupTransform()

    # 进入事件循环，处理定时器和TransformListener等回调
    rclpy.spin(node)

    # 销毁节点
    node.destroy_node()

    # 关闭ROS 2
    rclpy.shutdown()

File: 08_tf2_tutorial/src/lookup_transform_cpp/src/lookup_transform.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"

#include "geometry_msgs/msg/transform_stamped.hpp"

#include "tf2/exceptions.hpp"
#include "tf2/time.hpp"

#include "tf2_ros/buffer.hpp"
#include "tf2_ros/transform_listener.hpp"

using namespace std::chrono_literals;

class LookupTransform : public rclcpp::Node
{
public:
  LookupTransform()
  : Node("lookup_transform_cpp")
  {
    // 创建Buffer
    tf_buffer_ = std::make_shared<tf2_ros::Buffer>(
      this->get_clock()
    );

    // 创建TransformListener
    tf_listener_ =
      std::make_shared<tf2_ros::TransformListener>(
        *tf_buffer_,
        this,
        false
      );

    // 创建定时器
    // 每1s查询一次Transform，即查询频率约为1 Hz
    timer_ = this->create_timer(
      1s,
      std::bind(&LookupTransform::timer_callback, this)
    );

    // 打印启动日志
    RCLCPP_INFO(
      this->get_logger(),
      "开始查询Transform"
    );
  }

private:
  void lookup_latest_transform(
    const std::string & target,
    const std::string & source)
  {
    try {
      // 查询最新可用的Transform
      // target_frame为target
      // source_frame为source
      auto transform = tf_buffer_->lookupTransform(
        target,
        source,
        tf2::TimePointZero
      );

      // 获取平移分量
      const auto & translation =
        transform.transform.translation;

      // 获取旋转分量
      const auto & rotation =
        transform.transform.rotation;

      // 打印查询结果
      RCLCPP_INFO(
        this->get_logger(),
        "查询到 %s 转换到 %s Transform: "
        "translation=[%.3f, %.3f, %.3f], "
        "rotation=[%.3f, %.3f, %.3f, %.3f]",
        source.c_str(),
        target.c_str(),
        translation.x,
        translation.y,
        translation.z,
        rotation.x,
        rotation.y,
        rotation.z,
        rotation.w
      );
    }
    catch (const tf2::TransformException & ex) {
      // 当前Buffer中不存在可用Transform时打印警告
      RCLCPP_WARN(
        this->get_logger(),
        "暂时无法查询 %s 转换到 %s 的 Transform: %s",
        source.c_str(),
        target.c_str(),
        ex.what()
      );
    }
  }

  void timer_callback()
  {
    RCLCPP_INFO(
      this->get_logger(),
      "===================="
    );

    // 查询base_link(source)转换到odom(target)的Transform
    lookup_latest_transform(
      "odom",
      "base_link"
    );

    // 查询odom转换到base_link的Transform
    // Buffer帮我们完成逆变换
    lookup_latest_transform(
      "base_link",
      "odom"
    );

    // 查询base_link转换到map的Transform
    // Buffer帮我们完成复合变换
    lookup_latest_transform(
      "map",
      "base_link"
    );

    RCLCPP_INFO(
      this->get_logger(),
      "===================="
    );
  }

  // Buffer
  std::shared_ptr<tf2_ros::Buffer> tf_buffer_;

  // TransformListener
  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;

  // 定时器
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
  // 初始化ROS 2
  rclcpp::init(argc, argv);

  // 创建节点并进入事件循环
  rclcpp::spin(
    std::make_shared<LookupTransform>()
  );

  // 关闭ROS 2
  rclcpp::shutdown();

  return 0;
}
